opal/mca/btl/self/btl_self.c

```diff
@@ -218,7 +218,7 @@ static int mca_btl_self_send(struct mca_btl_base_module_t *btl,
     if (btl_ownership) {
         mca_btl_self_free(btl, des);
     }
-    return 1;
+    return OPAL_SUCCESS;
```

Comment:
There is a valid reason to return 1 from this async function. It is to indicate that everything has been taken care of, and the message is already done despite the fact that we expect it to happen in an async way. Without this return the upper level cannot make assumptions about the code in the case where MCA_BTL_DES_SEND_ALWAYS_CALLBACK is not provided.

Reply:
This is a byproduct of btl/base now calling btl_send() directly. Here it expects a return code of OPAL_SUCCESS, or else it does not remove the message from the queue. For btl/self this is problematic, and will result in a double free:

https://github.com/open-mpi/ompi/blob/master/opal/mca/btl/base/btl_base_am_rdma.c#L784

Reply:
I'm not sure of the best way to address this, maybe it should always remove it from the queue?

Reply:
The code you pointed to is incorrect, it should not test for OMPI_SUCCESS but should use `OPAL_UNLIKELY(rc >= 0)` as everywhere else. I think the fix is simple, you can remove the item from the list as soon as the ret is `OPAL_UNLIKELY(ret >= 0)`, as this means either that the message has been already sent, or that there will be a callback once the send completes (but in both cases the BTL will handle the rest of this message lifetime).

Reply:
@hjelmn any comment here.

Reply:
@bosilca Yes. We should be checking if the return code is >= 0 not just success. Forgot about that case. 1 means the data is gone.

Reply:
Making this change back to returning 1 for btl_self_send() seems to open up a whole can of worms for osc/rdma that need to get flushed out. Unfortunately it doesn't seem like a trivial fix - it just exposes other incompatibilities.